### ISSUE TITLE
Support using ICU when running on Windows 

### ISSUE THREAD

**Issue by tarekgh** [dotnet/corefx#41663]
Windows 10 now carrying a version of ICU. It will be good to allow .NET Core take advantage of that.

**Comment by danmosemsft** [comment]
I believe this also tracks potentially carrying a private version of ICU, whether on Linux or Windows. (We can split that out of course)

**Comment by tarekgh** [comment]
Right. This issue intended to track all related ICU work we need to do. we can add all information here.

**Comment by danmosemsft** [comment]
You mention supporting app-local ICU in Windows and Linux. Will we support it in macOS?

Are there any special behavior in Alpine here? I know it originally motivated the "no ICU" mode -- I assume ICU can be deployed on Alpine -- is it possible to use an app-local ICU there as well?

**Comment by tarekgh** [comment]
> You mention supporting app-local ICU in Windows and Linux. Will we support it in macOS?

Yes, macOS should be covered too. I'll edit the text to clarify one thing for supporting ICU as local app on Linux/macOS which is, the ICU NuGet packages which going to be produced by Windows team is going to support Windows only. So, on Linux/macOS, the app will need manually include the ICU libs if want to have local support. That till we'll have some source producing ICU NuGet packages for Linux/macOS.

> Are there any special caveats for Alpine here (or containers)? I know Alpine originally motivated the "no ICU" mode -- I assume ICU can be deployed on Alpine manually -- will it be possible to use an app-local ICU there as well?

If the globalization invariant mode is enabled on Alpine, then we'll not load ICU at all (as we do today). 
I believe there is a way to install ICU on Alpine as mentioned in https://pkgs.alpinelinux.org/package/edge/main/x86/icu

**Comment by jkotas** [comment]
It is ok to have performance regressions if they are understood and done intentionally for a good reason. We have done this trade-off many times in the history of .NET.

**Comment by GrabYourPitchforks** [comment]
We also have a few options available to address any performance losses if they show up in hot paths. For example, we already have a few places where we special-case `OrdinalIgnoreCase` comparison and go down a fast path if the input is all-ASCII. We could investigate expanding the number of places which receive such an optimization. We could cache commonly-used cultures at the ICU interop layer. We could even contribute changes back to ICU if we identify ways to make the algorithms themselves better.

**Comment by iSazonov** [comment]
From my experiments with Simple Case Folding a two-level table is as fast as ASCII workarounds. I believe MSFT experts can make it even faster.

**Comment by adamsitnik** [comment]
> We'll use ICU by default as this is the future direction and we want to get users switch to that. Of course this will be kind of breaking change especially for sorting behavior and locale data.

I think that we should revisit all known platform differences and group them into 3 categories:

- ICU bugs (and report|fix them)
- invalid ICU API usage (requires fix on our side)
- by design (and document these differences)

An example of ICU bug that we've already fixed: https://github.com/unicode-org/icu/pull/840

Examples of things that can be bugs on our side or ICU side:

```cs
Console.WriteLine(CultureInfo.InvariantCulture.CompareInfo.IsPrefix("''b", "b", CompareOptions.IgnoreSymbols)); // false
Console.WriteLine(CultureInfo.InvariantCulture.CompareInfo.IsPrefix("a''b", "ab", CompareOptions.IgnoreSymbols)); // true
```

```cs
CompareInfo frenchCompare = new CultureInfo("fr-FR").CompareInfo;
Console.WriteLine(frenchCompare.Compare("\u0153", "oe")); // 1, should be 0
```

**Comment by adamsitnik** [comment]
> We could investigate expanding the number of places which receive such an optimization.

Personally I am not a big fan of such optimizations. It adds a LOT of complexity on our side.

> We could cache commonly-used cultures at the ICU interop layer. 

I am not sure if I understand correctly, but we already have a cache:

https://github.com/dotnet/coreclr/blob/c6697a00ef7ecba6012b1a52f13401ce06b8295b/src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs#L42

> We could even contribute changes back to ICU if we identify ways to make the algorithms themselves better.

I had the "pleasure" to get more familiar with ICU code and I can say that there is definitely a place for improvement. The lack of possibility of breaking a 20 years old API makes it a little bit harder, but not impossible.

An example:

before we call the equivalent of `IndexOf` or `LastIndexOf` we need to call `usearch_openFromCollator` which calls method `initialize`, which calls `initializePattern` method which calls `initializePatternCETable`

`initializePatternCETable` allocates an array of integers of length of the "pattern" string and translates every character in the pattern string to an int32 (collation element). The problem is that it does it for every input, even for things like `"a".IndexOf(new string('a', 1_000_000))` It's OK for most of the use cases, the point that I am trying to make here is that the API that we need the most (collators) does not seem to be optimized for perf.

@tarekgh are there other parties at Microsoft interested in using ICU in performance-sensitive code? Maybe we should just identify the perf bottlenecks, submit proposals and somehow divide the work amongst teams. Also, if we really want to use ICU as our default we should ask for expanding the `usearch` API with `StartsWith` and `EndsWith` methods

**Comment by tarekgh** [comment]
> are there other parties at Microsoft interested in using ICU in performance-sensitive code?

We got some parties asking for using ICU on Windows. They already used it on Linux and never heard any complain from them about the perf there. So far, we didn't get anyone talked about the expected performance at all.

**Comment by jkotas** [comment]
It is super rare to see globalization APIs used on performance critical paths. If somebody uses globalization API on a performance critical path, it is most likely a bug - a wrong API being called.

**Comment by adamsitnik** [comment]
> It is super rare to see globalization APIs used on performance critical paths. 

I agree. The problem is that `StringComparison.CurrentCulture` is our default for all string based operations. So anytime our customers call `string.IndexOf`, `string.LastIndexOf`, `string.StartsWith`, `string.EndsWith` without specifying the `StringComparison` in explicit way, they hit the "slow path". Most software engineers don't measure the performance regulary, so typically they realize that something is "hot" after they hit a problem in produciton. In my opinion, we should try to prevent that from happening.

I agree with @kevingosse who wrote in https://github.com/dotnet/corefx/issues/40674#issuecomment-526319388

> The problem is that developers on .net framework or .net core/windows aren't really used to adding StringComparer.Ordinal, since the performance of culture-based string comparisons is pretty good on Windows (so it's mostly seen as a micro optimization). And then it breaks when we start using those bits on Linux.

I believe that if we really want to switch to ICU by default without optimizing it first, we should consider adding a `Ordinal` global mode, similar to global invariant mode we have.

**Comment by jkotas** [comment]
> anytime our customers call string.IndexOf, string.LastIndexOf, string.StartsWith, string.EndsWith without specifying the StringComparison in explicit way, they hit the "slow path"

They will hit the "slow path" only if the strings do contain globalization-sensitive characters.

> we should consider adding a Ordinal global mode, similar to global invariant mode we have.

Part of the plan is to have a switch to use NLS. So if somebody hits a problem with ICU (whether it is functionality or performance), they can go back to the legacy NLS to workaround it.

**Comment by jkotas** [comment]
> StringComparer.Ordinal, since the performance of culture-based string comparisons is pretty good on Windows (so it's mostly seen as a micro optimization)

Using StringComparer.Ordinal properly is not just a performance micro optimization. It is often required for security.

**Comment by adamsitnik** [comment]
> They will hit the "slow path" only if the strings do contain globalization-sensitive characters.

This is true only for `en*` cultures. The "slow path" is the default for every other culture.

**Comment by kevingosse** [comment]
> I believe that if we really want to switch to ICU by default without optimizing it first, we should consider adding a `Ordinal` global mode, similar to global invariant mode we have.

I'm a big supporter of this. It is really weird to say on one hand "people should use Ordinal most of the time anyway" and yet have APIs use culture-sensitive comparison per default (thus requiring extra knowledge to do the right thing).
